crdt.go

```diff
@@ -309,91 +357,162 @@ func (store *Datastore) handleBlock(ctx context.Context, data []byte) error {
 	}
 
 	// Walk down from this block.
-	err = store.walkBranch(c, c, 1)
-	if err != nil {
-		return errors.Wrap(err, "error walking new branch")
+	ctx, cancel := context.WithCancel(store.ctx)
+	defer cancel()
+
+	dg := &crdtNodeGetter{NodeGetter: store.dagSyncer}
+	if sessionMaker, ok := store.dagSyncer.(SessionDAGSyncer); ok {
+		dg = &crdtNodeGetter{NodeGetter: sessionMaker.Session(ctx)}
 	}
+
+	var session sync.WaitGroup
+	store.sendNewJobs(&session, dg, c, 0, []cid.Cid{c})
+	session.Wait()
 	return nil
 }
 
-// walkBranch walks down a branch and applies each node's delta until it
-// arrives to a known head or the bottom.  The given CID is assumed to not be
-// a known block and will be fetched.
-func (store *Datastore) walkBranch(current, top cid.Cid, depth uint64) error {
-	//store.logger.Debugf("walking on %s, from %s, depth %d", current, top, depth)
+// This runs in gorountine
+func (store *Datastore) dagWorker() {
+	for job := range store.jobQueue {
+		children, err := store.processNode(
+			job.nodeGetter,
+			job.root,
+			job.rootPrio,
+			job.delta,
+			job.node,
+		)
 
-	// TODO: Pre-fetching of children?
-	nd, delta, err := store.dags.GetDelta(store.ctx, current)
-	if err != nil {
-		return errors.Wrapf(err, "error fetching block %s", current)
+		if err != nil {
+			store.logger.Error(err)
+			job.session.Done()
+			continue
+		}
+		go func(j *dagJob) {
```

Comment:
it is,

```
// Close shuts down the CRDT datastore. It should not be used afterwards.
func (store *Datastore) Close() error {
	store.cancel()
	store.wg.Wait()
	return nil
}
```

Close cancels the context, which cancels the sendJobWorker:

```go
	for {
		select {
		case <-store.ctx.Done():
			close(store.jobQueue)
			return
		case j := <-store.sendJobs:
			store.jobQueue <- j
		}
	}
```

which closes the job queue which causes jobs to exit. On the other side, `sendNewJobs` goroutines also cancel because the context is cancelled and they cannot send more jobs because the sendJobWorker is not consuming the channel anymore.